README.md
# adsb-hassio-sensors

  Aircraft sensors for Home Assistant [adsb-multi-portal-feeder [from 1.26.1] ](https://github.com/MaxWinterstein/homeassistant-addons/tree/main/adsb-multi-portal-feeder) add-on by [MaxWinterstein](https://github.com/MaxWinterstein)
  
  

## Supported feeders

  adsb.fi

  adsbexchange.com

  flightradar24.com

  opensky-network.org

  planefinder.net


## Available sensors

| Feeder                | Friendly name                    | Sensor                          | Unit           |
| --------------------- | -------------------------------- | ------------------------------- | -------------- |
| `adsb.fi`             | `Adsb.fi Uptime`                 | sensor.adsbfi_uptime            |                |
| `adsb.fi`             | `Adsb.fi Status`                 | sensor.adsbfi_status            | aircraft       |
| `adsb.fi`             | `Adsb.fi ADS-B`                  | sensor.adsbfi_adsb              | aircraft       |
| `adsb.fi`             | `Adsb.fi Mode-S`                 | sensor.adsbfi_mode_s            | aircraft       |
| `adsb.fi`             | `Adsb.fi MLAT`                   | sensor.adsbfi_mlat              | aircraft       |
| `adsb.fi`             | `Adsb.fi Tracks`                 | sensor.adsbfi_tracks_all        | track          |
| `adsb.fi`             | `Adsb.fi Max Range`              | sensor.adsbfi_max_range         | nm             |

| Feeder                | Friendly name                    | Sensor                          | Unit           |
| --------------------- | -------------------------------- | ------------------------------- | -------------- |
| `adsbexchange.com`    | `Adsb Exchange Uptime`           | sensor.adsbx_uptime             |                |
| `adsbexchange.com`    | `Adsb Exchange Status`           | sensor.adsbx_status             | aircraft       |
| `adsbexchange.com`    | `Adsb Exchange ADS-B`            | sensor.adsbx_adsb               | aircraft       |
| `adsbexchange.com`    | `Adsb Exchange Mode-S`           | sensor.adsbx_mode_s             | aircraft       |
| `adsbexchange.com`    | `Adsb Exchange MLAT`             | sensor.adsbx_mlat               | aircraft       |
| `adsbexchange.com`    | `Adsb Exchange Tracks`           | sensor.adsbx_tracks_all         | track          |
| `adsbexchange.com`    | `Adsb Exchange Max Range`        | sensor.adsbx_max_range          | nm             |

| Feeder                | Friendly name                    | Sensor                          | Unit           |
| --------------------- | -------------------------------- | ------------------------------- | -------------- |
| `flightradar24.com`   | `Flightradar24 Feeder Status`    | sensor.flightradar_feed         |                |
| `flightradar24.com`   | `Flightradar24 Status`           | sensor.flightradar_status       | aircraft       |
| `flightradar24.com`   | `Flightradar24 ADS-B`            | sensor.flightradar_adsb         | aircraft       |
| `flightradar24.com`   | `Flightradar24 Non ADS-B`        | sensor.flightradar_non_adsb     | aircraft       |
| `flightradar24.com`   | `Flightradar24 Max Range`        | sensor.flightradar_max_range    | nm             |

| Feeder                | Friendly name                    | Sensor                          | Unit           |
| --------------------- | -------------------------------- | ------------------------------- | -------------- |
| `opensky-network.org` | `Opensky Status`                 | sensor.opensky_status           | aircraft       | 
| `opensky-network.org` | `Opensky ADS-B`                  | sensor.opensky_adsb             | aircraft       | 
| `opensky-network.org` | `Opensky MLAT`                   | sensor.opensky_mlat             | aircraft       |
| `opensky-network.org` | `Opensky Mode-S Rate`            | sensor.opensky_mode_s_rate      | aircraft / min | 
| `opensky-network.org` | `Opensky Mode-AC Rate`           | sensor.opensky_mode_ac_rate     | aircraft / min | 
| `opensky-network.org` | `Opensky Max Range`              | sensor.opensky_max_range        | nm             | 

| Feeder                | Friendly name                    | Sensor                          | Unit           |
| --------------------- | -------------------------------- | ------------------------------- | -------------- |
| `planefinder.net`     | `Planefinder Status`             | sensor.planefinder_status       | aircraft       |
| `planefinder.net`     | `Planefinder ADS-B`              | sensor.planefinder_adsb         | aircraft       |
| `planefinder.net`     | `Planefinder MLAT`               | sensor.planefinder_mlat         | aircraft       |
| `planefinder.net`     | `Planefinder Mode-S Rate`        | sensor.planefinder_mode_s_rate  | packet / s     |
| `planefinder.net`     | `Planefinder Mode-AC Rate`       | sensor.planefinder_mode_ac_rate | packet / s     |
| `planefinder.net`     | `Planefinder Receiver Bandwidth` | sensor.planefinder_bandwidth    | Kb / s         |
| `planefinder.net`     | `Planefinder Max Range`          | sensor.planefinder_max_range    | nm             |


## Configuration

  `Max Range` sensors require `MLAT_EXACT_LAT` `MLAT_EXACT_LON` configured

##

![sensor aircraft tracked](https://github.com/plo53/adsb-hassio-sensors/blob/master/media/Home%20Assistant%20ADS-B%20sensors.png)



CHANGELOG.md
## [1.1.2] - 2023-07-16

- New Feature : `Flightradar24` feeder support
- New Feature : `Uptime` sensor for `adsb.fi` `Adsb Exchange` feeders
- New Feature : `Max Range` sensor for all feeders

## [1.1.1] - 2023-06-27

- Updates : New total Track sensor for `adsb.fi` , `Adsb Exchange`

## [1.1.0] - 2023-06-26

- New Feature : `Opensky` feeder support
- New Feature : `Planefinder` feeder support
- Updates : Long-term Statistics for all sensors
- Fix : Sensor unique_id issue

## [1.0.1] - 2023-06-23

- Compatibility Fix to support `MaxWinterstein/adsb-multi-portal-feeder` integration

## [1.0.0] - 2023-06-23

- Initial release


usr/local/share/adsbfi/adsbfi_max-range.py
#!/usr/bin/python3

# Max range from geodesic aircraft distance (adsb.fi)

from urllib.request import urlopen
import json
from math import *


def distance(lat1,lon1,lat2,lon2):

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    d = 2 * atan2(sqrt(a), sqrt(1 - a))

    return d

if __name__ == "__main__":

    R = 6373.0 / 1.852

    with open('/data/options.json', 'r') as f:
        options = json.load(f)
        if 'MLAT_EXACT_LAT' in options:
            ulat=radians(float( options['MLAT_EXACT_LAT'] ))
        else:
            ulat=float(0)
        if 'MLAT_EXACT_LON' in options:
            ulon=radians(float( options['MLAT_EXACT_LON'] ))
        else:
            ulon=float(0)

    url =  '/run/adsbfi-feed/aircraft.json'

    with open(url) as response:

        data = json.loads(response.read())

        dist = {}

        for values in data['aircraft']:
                icao=values.get('hex')
                lat=values.get('lat')
                lon=values.get('lon')
                if (lat is not None) and (lon is not None):
                    lat=radians(float(lat))
                    lon=radians(float(lon))
                    dist[icao] = distance(ulat,ulon,lat,lon) * R

        sorted_dist=sorted(dist.items(), key=lambda x:x[1], reverse=True)

        if (sorted_dist and ulat != 0 and ulon != 0):
            sensor_mr = {"state": int(sorted_dist[0][1]),
                         "attributes": {"friendly_name": "Adsb.fi Max Range",
                         "icon": "mdi:map-marker-distance",
                         "state_class": "measurement",
                         "unit_of_measurement": "nm",
                         "unique_id": "adsbfi_max_range"} }
        else:
            sensor_mr = {"state": "0",
                         "attributes": {"friendly_name": "Adsb.fi Max Range",
                         "icon": "mdi:map-marker-distance",
                         "state_class": "measurement",
                         "unit_of_measurement": "nm",
                         "unique_id": "adsbfi_max_range"} }

        print(json.dumps(sensor_mr))


usr/local/share/opensky/opensky_max-range.py
#!/usr/bin/python3

# Max range from geodesic aircraft distance (Opensky)
# from http://localhost:8080/data/aircraft.json

from urllib.request import urlopen
import json
from math import *


def distance(lat1,lon1,lat2,lon2):

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    d = 2 * atan2(sqrt(a), sqrt(1 - a))

    return d

if __name__ == "__main__":

    R = 6373.0 / 1.852

    with open('/data/options.json', 'r') as f:
        options = json.load(f)
        if 'MLAT_EXACT_LAT' in options:
            ulat=radians(float( options['MLAT_EXACT_LAT'] ))
        else:
            ulat=float(0)
        if 'MLAT_EXACT_LON' in options:
            ulon=radians(float( options['MLAT_EXACT_LON'] ))
        else:
            ulon=float(0)

    url =  'http://localhost:8080/data/aircraft.json'

    with urlopen(url) as response:

        data = json.loads(response.read())

        dist = {}

        for values in data['aircraft']:
                icao=values.get('hex')
                lat=values.get('lat')
                lon=values.get('lon')
                if (lat is not None) and (lon is not None):
                    lat=radians(float(lat))
                    lon=radians(float(lon))
                    dist[icao] = distance(ulat,ulon,lat,lon) * R

        sorted_dist=sorted(dist.items(), key=lambda x:x[1], reverse=True)

        if (sorted_dist and ulat != 0 and ulon != 0):
            sensor_mr = {"state": int(sorted_dist[0][1]),
                     	"attributes": {"friendly_name": "Opensky Max Range",
                     	"icon": "mdi:map-marker-distance",
                     	"state_class": "measurement",
                     	"unit_of_measurement": "nm",
                     	"unique_id": "opensky_max_range"} }
        else:
            sensor_mr = {"state": "0",
                     	"attributes": {"friendly_name": "Opensky Max Range",
                     	"icon": "mdi:map-marker-distance",
                     	"state_class": "measurement",
                     	"unit_of_measurement": "nm",
                     	"unique_id": "opensky_max_range"} }

        print(json.dumps(sensor_mr))


usr/local/share/flightradar/fr24_max-range.py
#!/usr/bin/python3

# Max range from geodesic aircraft distance (FR24)
# from localhost:8754/flights.json

from urllib.request import urlopen
import json
from math import *


def distance(lat1,lon1,lat2,lon2):

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    d = 2 * atan2(sqrt(a), sqrt(1 - a))

    return d

if __name__ == "__main__":

    R = 6373.0 / 1.852

    with open('/data/options.json', 'r') as f:
        options = json.load(f)
        if 'MLAT_EXACT_LAT' in options:
            ulat=radians(float( options['MLAT_EXACT_LAT'] ))
        else:
            ulat=float(0)
        if 'MLAT_EXACT_LON' in options:
            ulon=radians(float( options['MLAT_EXACT_LON'] ))
        else:
            ulon=float(0)

    url =  'http://localhost:8754/flights.json'

    with urlopen(url) as response:

        data = json.loads(response.read())

        dist = {}

        for key in data:
             icao=data[key][0]
             lat=data[key][1]
             lon=data[key][2]
             if (lat != 0.) and (lon != 0.):
                 lat=radians(float(lat))
                 lon=radians(float(lon))
                 dist[icao] = distance(ulat,ulon,lat,lon) * R

        sorted_dist=sorted(dist.items(), key=lambda x:x[1], reverse=True)

        if (sorted_dist and ulat != 0 and ulon != 0):
            sensor_mr = {"state": int(sorted_dist[0][1]),
                      "attributes": {"friendly_name": "Flightradar24 Max Range",
                      "icon": "mdi:map-marker-distance",
                      "state_class": "measurement",
                      "unit_of_measurement": "nm",
                      "unique_id": "fr24_max_range"} }
        else:
            sensor_mr = {"state": "0",
                      "attributes": {"friendly_name": "Flightradar24 Max Range",
                      "icon": "mdi:map-marker-distance",
                      "state_class": "measurement",
                      "unit_of_measurement": "nm",
                      "unique_id": "fr24_max_range"} }

        print(json.dumps(sensor_mr))


usr/local/share/adsbexchange/adsbx_max-range.py
#!/usr/bin/python3

# Max range from geodesic aircraft distance (Adsb Exchange)

from urllib.request import urlopen
import json
from math import *


def distance(lat1,lon1,lat2,lon2):

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    d = 2 * atan2(sqrt(a), sqrt(1 - a))

    return d

if __name__ == "__main__":

    R = 6373.0 / 1.852

    with open('/data/options.json', 'r') as f:
        options = json.load(f)
        if 'MLAT_EXACT_LAT' in options:
            ulat=radians(float( options['MLAT_EXACT_LAT'] ))
        else:
            ulat=float(0)
        if 'MLAT_EXACT_LON' in options:
            ulon=radians(float( options['MLAT_EXACT_LON'] ))
        else:
            ulon=float(0)

    url =  '/run/adsbexchange-feed/aircraft.json'

    with open(url) as response:

        data = json.loads(response.read())

        dist = {}

        for values in data['aircraft']:
                icao=values.get('hex')
                lat=values.get('lat')
                lon=values.get('lon')
                if (lat is not None) and (lon is not None):
                    lat=radians(float(lat))
                    lon=radians(float(lon))
                    dist[icao] = distance(ulat,ulon,lat,lon) * R

        sorted_dist=sorted(dist.items(), key=lambda x:x[1], reverse=True)

        if (sorted_dist and ulat != 0 and ulon != 0):
            sensor_mr = {"state": int(sorted_dist[0][1]),
                         "attributes": {"friendly_name": "Adsb Exchange Max Range",
                         "icon": "mdi:map-marker-distance",
                         "state_class": "measurement",
                         "unit_of_measurement": "nm",
                         "unique_id": "adsbx_max_range"} }
        else:
            sensor_mr = {"state": "0",
                         "attributes": {"friendly_name": "Adsb Exchange Max Range",
                         "icon": "mdi:map-marker-distance",
                         "state_class": "measurement",
                         "unit_of_measurement": "nm",
                         "unique_id": "adsbx_max_range"} }

        print(json.dumps(sensor_mr))


usr/local/share/planefinder/planefinder_max-range.py
#!/usr/bin/python3

# Max range from geodesic aircraft distance (PlaneFonder)
# from http://localhost:30053/ajax/aircraft

from urllib.request import urlopen
import json
from math import *


def distance(lat1,lon1,lat2,lon2):

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    d = 2 * atan2(sqrt(a), sqrt(1 - a))

    return d

if __name__ == "__main__":

    R = 6373.0 / 1.852

    with open('/data/options.json', 'r') as f:
        options = json.load(f)
        if 'MLAT_EXACT_LAT' in options:
            ulat=radians(float( options['MLAT_EXACT_LAT'] ))
        else:
            ulat=float(0)
        if 'MLAT_EXACT_LON' in options:
            ulon=radians(float( options['MLAT_EXACT_LON'] ))
        else:
            ulon=float(0)

    url =  'http://localhost:30053/ajax/aircraft'

    with urlopen(url) as response:

        data = json.loads(response.read())

        dist = {}

        for key, values in data['aircraft'].items():
            if values['is_mlat'] != True:
                lat=values.get('lat')
                lon=values.get('lon')
                if (lat is not None) and (lon is not None):
                    lat=radians(float(lat))
                    lon=radians(float(lon))
                    dist[key] = distance(ulat,ulon,lat,lon) * R

        sorted_dist=sorted(dist.items(), key=lambda x:x[1], reverse=True)

        if (sorted_dist and ulat != 0 and ulon != 0):
            sensor_mr = {"state": int(sorted_dist[0][1]),
                         "attributes": {"friendly_name": "Planefinder Max Range",
                         "icon": "mdi:map-marker-distance",
                         "state_class": "measurement",
                         "unit_of_measurement": "nm",
                         "unique_id": "planefinder_max_range"} }
        else:
            sensor_mr = {"state": "0",
                         "attributes": {"friendly_name": "Planefinder Max Range",
                         "icon": "mdi:map-marker-distance",
                         "state_class": "measurement",
                         "unit_of_measurement": "nm",
                         "unique_id": "planefinder_max_range"} }

        print(json.dumps(sensor_mr))
